### PR TITLE
Add volk_32f(c)_index_min_16/32u

### LINKED ISSUE #481: Add  volk_32f(c)_index_min_16u/32u
Hi all,
It would be nice to have volk_32f(c)_index_min_16u/32u functions in addition to the existing volk_32f(c)_index_max_16u/32u functions. Can I try to propose a PR to add these functions?
Regards,
Thomas

### PR DESCRIPTION
**Pull request by Zlika**
Added functions volk_32f(c)_index_min_16/32u based on the already existing functions volk_32f(c)_index_max_16/32u.
Fixes #481.
Signed-off-by: Zlika <zlika_ese@hotmail.com>

### CODE REVIEW

**Comment by marcusmueller** [comment]
Hi Zlika! Since you've got "(wip)" in your question title, I thought it'd be wise to set this PR to "draft" mode (which has no other consequences than telling the maintainers to not merge it yet). You can remove the draft mode by clicking on "Ready for review" at any time!

**Comment by Zlika** [comment]
Hi @marcusmueller. Thanks for your tips, I wasn't aware of github's draft feature for PR. Now this PR is ready for review. I don't think the windows build failure is related to this PR.

File: kernels/volk/volk_32f_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32f_index_min_16u.h]: I suggest to update the parameter name `src0`. There's no `src1` anyways. What about `src`? Or probably better `source`. 
Since this is a search in a list or array, you might want to call it `list` or `array`.

File: kernels/volk/volk_32f_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32f_index_min_16u.h]: I suggest to make variables as local as possible. Could you move `i` into the for-loop, i.e. `for(uint32_t i = 1; ...`.
I would suggest that approach for all loops. After all, it makes code easier to read.

File: kernels/volk/volk_32f_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32f_index_min_16u.h:299]: We need to double check if this is still necessary. From the top of my head, I'd say no, we can just use one include guard per file with all the includes at the top.

File: kernels/volk/volk_32f_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32f_index_min_16u.h:3]: We switched to DCO, so you don't need to assign copyright to the FSF anymore. Furthermore, did you sign a CLA with the FSF in the past? If yes, you can still assign copyright if you want to, if no, you shouldn't.

File: kernels/volk/volk_32f_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32f_index_min_16u.h]: Into the for-loop please.

File: kernels/volk/volk_32f_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32f_index_min_16u.h]: Into the for-loop please.

File: kernels/volk/volk_32f_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32f_index_min_16u.h]: Into the for-loop please.

File: kernels/volk/volk_32f_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32f_index_min_16u.h]: Into the for-loop please.

File: kernels/volk/volk_32f_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32f_index_min_16u.h]: Into the for-loop please.

File: kernels/volk/volk_32f_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32f_index_min_16u.h]: Into the for-loop please.

File: kernels/volk/volk_32f_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32f_index_min_16u.h]: Into the for-loop please.

File: kernels/volk/volk_32f_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32f_index_min_16u.h]: Into the for-loop please.

**Review by jdemel — CHANGES_REQUESTED** [review]
Thanks for your PR. I added comments to the first kernel. Mostly, the code looks pretty good. But there are a few generic things that need to be updated. I assume the same comments apply to the other kernels as well.

**Comment by Zlika** [comment]
Thanks for your remarks @jdemel. I agree with all of them. I didn't "clean" the code because this is mostly a copy/paste of the existing functions volk_32f(c)_index_max_16/32u, and so I thought there might be some optimization black magic with these strange "off the loop" variable assignments.
If I clean up this code, it would be nice to do the same for the existing functions volk_32f(c)_index_max_16/32u as they share 99% of the code.
By the way, I signed the FSF CLA one or two years ago.

**Comment by jdemel** [comment]
I totally agree. It'd be great to do some clean-up in those kernels as well.

File: kernels/volk/volk_32f_index_min_16u.h
Comment (Zlika) [reply kernels/volk/volk_32f_index_min_16u.h:299]: I tried to use a single include guard, but in this case the tests only run the "aligned" versions, and the "best unaligned arch" in the tests is always "generic". I don't know how it works under the hood, but it seems that these two include guards are needed.

File: kernels/volk/volk_32f_index_min_16u.h
Comment (Zlika) [reply kernels/volk/volk_32f_index_min_16u.h:3]: I signed the CLA two years ago (June 2019).

**Comment by Zlika** [comment]
Hi @jdemel 
I took into account all your remarks, with the exception of the "double include guards" (unaligned kernels are not taken into account if I use a single "include guard", I don't know why) and the copyright (I signed the CLA).

File: kernels/volk/volk_32f_index_min_32u.h
Comment (jdemel) [inline kernels/volk/volk_32f_index_min_32u.h]: I would really like to know where this if-statement originates. I assume you adapted the corresponding `index_max` kernel but why would this kernel have this if-statement?
I'd recommend to remove it. Calling this kernel with `num_points == 0` basically implies we call this kernel with a NULL pointer. The `index` (or `target`) should point to the zeroth element in this case.

File: kernels/volk/volk_32fc_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32fc_index_min_16u.h]: I suggest to update this line in all files to:
```
 * This file is part of VOLK
```

File: kernels/volk/volk_32fc_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32fc_index_min_16u.h]: Could you remove the `num_bytes` variable from this kernel? It is only used here and only contributes to confusion. I'd suggest to use `num_points` in the loop.

File: kernels/volk/volk_32fc_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32fc_index_min_16u.h:246]: Would you be interested in creating new intrinsics in `include/volk/volk_sse3_intrinsics.h`? I assume these would be useful in another new kernel as well. Thus, we would reduce code copypasta.

File: kernels/volk/volk_32fc_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32fc_index_min_16u.h]: Could you remove this variable from this kernel? Basically, it is never used directly and could be replaced with `num_points` usage.

File: kernels/volk/volk_32fc_index_min_32u.h
Comment (jdemel) [inline kernels/volk/volk_32fc_index_min_32u.h]: I suggest to get rid of `num_bytes` in this kernel.

File: kernels/volk/volk_32fc_index_min_32u.h
Comment (jdemel) [inline kernels/volk/volk_32fc_index_min_32u.h:243]: This would be a good reason to move this to its own function in `volk_sse3_intrinsics.h`, cf. other kernel comment.

File: kernels/volk/volk_32fc_index_min_32u.h
Comment (jdemel) [inline kernels/volk/volk_32fc_index_min_32u.h]: I suggest to get rid of this variable.

File: kernels/volk/volk_32fc_index_min_32u.h
Comment (jdemel) [inline kernels/volk/volk_32fc_index_min_32u.h:474]: Same as other comments. Why check if the input is not completely nonsensical?

File: kernels/volk/volk_32fc_index_min_32u.h
Comment (jdemel) [inline kernels/volk/volk_32fc_index_min_32u.h]: I suggest to update the function signature to `const lv_32fc_t* source` to make it more obvious which parameters get updated.

File: kernels/volk/volk_32fc_index_min_16u.h
Comment (jdemel) [inline kernels/volk/volk_32fc_index_min_16u.h]: I suggest to update the function signature to const lv_32fc_t* source to make it more obvious which parameters get updated.

**Review by jdemel — CHANGES_REQUESTED** [review]
Thanks for your patience! I have a few more comments. I hope this changes can be implemented easily.

**Comment by Zlika** [comment]
Hi @jdemel . Thanks for your remarks. I took them all into account (except your proposition to create a new function in volk_sse3_intrinsics.h because it requires more thinking on my side).

**Review by jdemel — APPROVED** [review]
LGTM now. Thanks for your effort and patience.
Since we're in the process to relicense VOLK under LGPL, would you be willing to contribute under the new license as well? The quickest way to show your approval would be to "suggest a change" in #486 that adds your name and info to the "Table of authors".

**Comment by Zlika** [comment]
Hi @jdemel. Thanks for your help and guidance. I agree to contribute my work under the LGPL license, but as index_min is clearly a derived work of index_max, index_max code should also switch to LGPL.

**Comment by jdemel** [comment]
@Zlika Thanks!
We're working on setting up a process to ask every contributor.

**Review by michaelld — APPROVED** [review]
LGTM ; thx!

**Comment by michaelld** [comment]
@Zlika regarding the relicensing: let us merge #486 & then the easiest way forward is to add your info the noted file as a PR. We'll review & merge. In this manner we'll get the vast majority of VOLK relicensed in the next few weeks -- guessing through September given the momentum up to and following GRCon (Sep 20-24) & then the post GRCon "hangover". Once we hit some reasonable threshold and a given amount of time (TBD) has passed, then we'll remove any non-relicensed code & do a major release (3.0.0). We will then try to find ways to get the removed cost added back.

**Comment by Zlika** [comment]
Do I have to add my name to AUTHORS_RESUBMITTING_UNDER_LGPL_LICENSE.md before or after this PR is merged?

**Comment by michaelld** [comment]
> Do I have to add my name to AUTHORS_RESUBMITTING_UNDER_LGPL_LICENSE.md before or after this PR is merged?

Whatever works for you. Either is fine. Preference for just doing the relicensing now though.

**Comment by Zlika** [comment]
@michaelld I created PR https://github.com/gnuradio/volk/pull/512 to add my name. Thanks.

**Comment by jdemel** [comment]
We have 2 approvals and the whole LGPL business is finished. Merging.

Thanks!

**Pull request merged**
